sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -226,6 +258,41 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     result.copy(expectedOutputAttributes = Some(metastoreRelation.output))
   }
 
+  private def inferSchema(
+      metastoreSchema: StructType,
+      options: Map[String, String],
+      fileFormat: FileFormat,
+      fileType: String,
+      fileIndex: FileIndex): Option[StructType] = {
+    val inferred = fileFormat.inferSchema(
+      sparkSession,
+      options,
+      fileIndex.listFiles(Nil).flatMap(_.files))
+    if (fileType.equals("parquet")) {
+      inferred.map(ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, _))
+    } else {
+      inferred
+    }
+  }
+
+  private def updateCatalogTable(
+      catalogTable: CatalogTable,
+      inferredSchema: Option[StructType]): Option[CatalogTable] = try {
+    inferredSchema.flatMap { schema =>
+      logInfo(s"Saving case-sensitive schema for table ${catalogTable.identifier.table}")
+      val updatedTable = catalogTable.copy(schema = schema)
+      val catalog = sparkSession.sharedState.externalCatalog
+      catalog.alterTable(updatedTable)
+      Option(catalog.getTable(
```

Comment:
I think that should be fine. I had some concerns around the way ```HiveExternalCatalog``` mutates the raw ```CatalogTable``` returned by the metastore that I think pushed me towards fetching the table again but I really don't think that should matter since the original ```catalogTable``` was retrieved from ```HiveExternalCatalog``` as well.

I'll just used ```updatedTable``` here.